### ISSUE TITLE
`require-post-message-target-origin` shouldn't trigger for incompatible APIs

### ISSUE THREAD

**Issue by GauBen** [sindresorhus/eslint-plugin-unicorn#1396]
## What's wrong 
I'm currently developing a browser extension that communicates with native software:

```ts
import { browser, Runtime } from 'webextension-polyfill-ts'
const nativePort: Runtime.Port = browser.runtime.connectNative(APPLICATION_ID)
nativePort.postMessage({ cmd: 'Version' }) // (x) Missing the `targetOrigin` argument.
```

## Which rule is buggy
`require-post-message-target-origin` is triggered but the method only accepts one argument (see [runtime.Port](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/runtime/Port)).

## Ideas 

1. Enforce the name of the variable to be `window`, `document` or the likes.
2. Check the type of the variable or the existence of the import `'webextension-polyfill'`/`'webextension-polyfill-ts'`.
3. Add a warning for web extension developers here: https://github.com/sindresorhus/eslint-plugin-unicorn/blob/v34.0.1/docs/rules/require-post-message-target-origin.md

**Comment by fisker** [comment]
Sorry, didn't know this.

As I understand, `runtime.Port` don't have to come from `'webextension-polyfill'`/`'webextension-polyfill-ts'`, right?

**Comment by GauBen** [comment]
You understood right, `'webextension-polyfill'` (https://github.com/mozilla/webextension-polyfill) is just something that helps erasing differences between chromium and gecko. Especially, `chrome.runtime` becomes `browser.runtime`, as asked by the web extension standard.

**Comment by fisker** [comment]
Do you think disable this rule in your file or project is a good idea? If you are developing a browser extension, you probably don't need call window.postMessage.

**Comment by GauBen** [comment]
Yep, you're right, I shouldn't be using `window.postMessage` (https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage#using_window.postmessage_in_extensions_non-standard_inline)
I'll disable the rule in my eslintrc.

Thanks for the quick response!

**Comment by fisker** [comment]
But is it common to use worker in extension? In that case people may need `self.postMessage`?

**Comment by fisker** [comment]
Wait, [worker.postMessage](https://developer.mozilla.org/en-US/docs/Web/API/Worker/postMessage) second parameter is not `targetOrigin`....

**Comment by fisker** [comment]
There is also [MessagePort.postMessage()](https://developer.mozilla.org/en-US/docs/Web/API/MessagePort/postMessage)

**Comment by fisker** [comment]
I think we should turn this rule off...

**Comment by GauBen** [comment]
I'm not sure why this rule exists, the parameter `targetOrigin` is not optionnal, and the error is properly caught by tsc:

```
Expected 2-3 arguments, but got 1. ts(2554)
lib.dom.d.ts(18289, 31): An argument for 'targetOrigin' was not provided.
```

**Comment by fisker** [comment]
Our rule should support JavasScript, 

> When calling window.postMessage() without targetOrigin argument, browser won't throw any error, but message can't receive by any window. I remember the first time I got into this, really drive me crazy.

https://github.com/sindresorhus/eslint-plugin-unicorn/issues/1307#issue-897600197

**Comment by GauBen** [comment]
Ok got it, the problem is more that browsers fail silently instead of throwing an exception. What if this rule only triggers for `window.postMessage` by default in `unicorn/recommended`, and add an option to trigger on any variable like `foo.postMessage`?

```js
rules: {
  'unicorn/require-post-message-target-origin': ['error', 'window' | 'any'],
}
```

**Comment by JounQin** [comment]
This rule should not trigger on web worker, either.

```ts
const msg: MainToWorkerMessage = { sharedBuffer, id, args }
new Worker('', {}).postMessage(msg)
```

**Comment by fregante** [comment]
I think this rule is safe only if types are available:

- Does the `postMessage` function have a `targetOrigin` argument? Then report the error

Can that be done?

**Comment by fisker** [comment]
> Can that be done?

I don't think so, without type info, we can't know where the `postMessage` from.

**Comment by fregante** [comment]
> I think this rule is safe only if types are available:

👆 Yes I'm talking about a TypeScript-only rule.

**Comment by fisker** [comment]
That will be meaningless if this https://github.com/sindresorhus/eslint-plugin-unicorn/issues/1396#issuecomment-872931334 is true.

**Comment by fregante** [comment]
> That will be meaningless if this [#1396 (comment)](https://github.com/sindresorhus/eslint-plugin-unicorn/issues/1396#issuecomment-872931334) is true.

Indeed it is! So this rule can be dropped altogether as far as I'm concerned.

<img width="592" alt="Screen Shot 10" src="https://user-images.githubusercontent.com/1402241/125254759-f85b6880-e324-11eb-97bf-47fa04fd5df0.png">

**Comment by JounQin** [comment]
> That will be meaningless if this [#1396 (comment)](https://github.com/sindresorhus/eslint-plugin-unicorn/issues/1396#issuecomment-872931334) is true.

So `require-post-message-target-origin` should be disabled by default for `.ts` files in the recommended config.

**Comment by fisker** [comment]
Reopen this since `{Worker,MessagePort,Client,BroadcastChannel}#postMessage()` and `{browser,chrome}.postMessage()` still false positive in pure JS parsers.

**Comment by fisker** [comment]
@2000yeshu What's your solution for this?
